wfe/wfe.go

```diff
@@ -390,23 +399,116 @@ func (wfe *WebFrontEndImpl) NewRegistration(
 		return
 	}
 
-	wfe.db.addRegistration(&newReg)
-	wfe.log.Printf("There are now %d registrations in memory\n", wfe.db.countRegistrations())
+	count, err := wfe.db.addRegistration(&createdReg)
+	if err != nil {
+		wfe.sendError(acme.InternalErrorProblem("Error persisting registration"), response)
+		return
+	}
+	wfe.log.Printf("There are now %d registrations in memory\n", count)
 
-	regURL := wfe.relativeEndpoint(request, fmt.Sprintf("%s%s", regPath, newReg.ID))
+	regURL := wfe.relativeEndpoint(request, fmt.Sprintf("%s%s", regPath, regID))
 
 	response.Header().Add("Location", regURL)
 	err = wfe.writeJsonResponse(response, http.StatusCreated, newReg)
 	if err != nil {
 		wfe.sendError(acme.InternalErrorProblem("Error marshalling registration"), response)
+		return
+	}
+}
+
+func (wfe *WebFrontEndImpl) verifyOrder(order *core.Order, reg *core.Registration) *acme.ProblemDetails {
+	// Shouldn't happen - defensive check
+	if order == nil {
+		return acme.InternalErrorProblem("Order is nil")
+	}
+	if reg == nil {
+		return acme.InternalErrorProblem("Registration is nil")
+	}
+	csr := order.ParsedCSR
+	if csr == nil {
+		return acme.InternalErrorProblem("Parsed CSR is nil")
+	}
+	if len(csr.DNSNames) == 0 {
+		return acme.MalformedProblem("CSR has no names in it")
 	}
+	orderKeyID, err := keyToID(csr.PublicKey)
+	if err != nil {
+		return acme.MalformedProblem("CSR has an invalid PublicKey")
+	}
+	if orderKeyID == reg.ID {
+		return acme.MalformedProblem("Certificate public key must be different than account key")
+	}
+	return nil
 }
 
-func (wfe *WebFrontEndImpl) validateOrder(order *acme.Order) error {
-	// TODO(@cpu) - Apply some more advanced policy decisions on the CSR
+// makeAuthorizations populates an order with new authz's. The request parameter
+// is required to make the authz URL's absolute based on the request host
+func (wfe *WebFrontEndImpl) makeAuthorizations(order *core.Order, request *http.Request) error {
+	var auths []string
+
+	names := make([]string, len(order.ParsedCSR.DNSNames))
+	copy(names, order.ParsedCSR.DNSNames)
+
+	// Create one authz for each name in the CSR
+	for _, name := range names {
+		ident := acme.Identifier{
+			Type:  acme.IdentifierDNS,
+			Value: name,
+		}
+		authz := &core.Authorization{
+			ID: core.NewToken(),
+			Authorization: acme.Authorization{
+				Status:     acme.StatusPending,
+				Identifier: ident,
+			},
+		}
+		authz.URL = wfe.relativeEndpoint(request, fmt.Sprintf("%s%s", authzPath, authz.ID))
+		// Create the challenges for this authz
+		err := wfe.makeChallenges(authz, request)
+		if err != nil {
+			return err
+		}
+		// Persist the authorization in memory
+		count, err := wfe.db.addAuthorization(authz)
+		if err != nil {
+			return err
+		}
+		fmt.Printf("There are now %d authorizations in the db\n", count)
```

Comment:
I left these `printf`'s in on purpose to make it easier to discern the state of the Pebble server. I'll update them to use the logging instead of a raw printf. Fair?